Improve support of `mixed` type in objects

```zep
public function offsetGet(mixed index) -> mixed
{
    return this->{index};
}
```
generates error:
> [ERROR] Variable type: mixed cannot be used in property-dynamic-access in /srv/phalcon/Mvc/Model/Row.zep on line 43

Also
```zep
public function offsetExists(mixed index) -> bool
{
    return isset this->{index};
}
```
generates error:
> [ERROR] [mixed] in /srv/phalcon/Mvc/Model/Row.zep on line 55

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](http://semver.org).
 ## [Unreleased]
 ### Fixed
 - Fixed left `null` with `string` condition [#2299](https://github.com/zephir-lang/zephir/issues/2299)
+- Improved support of `mixed` type [#2330](https://github.com/zephir-lang/zephir/issues/2330)
 
 ## [0.15.2] - 2021-10-24
 ### Fixed

--- a/Library/Operators/Other/IssetOperator.php
+++ b/Library/Operators/Other/IssetOperator.php
@@ -13,6 +13,7 @@ declare(strict_types=1);
 
 namespace Zephir\Operators\Other;
 
+use ReflectionException;
 use Zephir\CompilationContext;
 use Zephir\CompiledExpression;
 use Zephir\Exception;
@@ -21,27 +22,24 @@ use Zephir\Expression;
 use Zephir\Operators\AbstractOperator;
 
 /**
- * Checks if a array offset or a property is defined on a polymorphic variable
+ * Checks if an array offset or a property is defined on a polymorphic variable
  */
 class IssetOperator extends AbstractOperator
 {
     /**
      * Compiles an 'isset' operator.
      *
-     * @param array              $expression
+     * @param array $expression
      * @param CompilationContext $compilationContext
      *
      * @return CompiledExpression
      *
      * @throws Exception
+     * @throws ReflectionException
      */
     public function compile(array $expression, CompilationContext $compilationContext): CompiledExpression
     {
-        if ('list' === $expression['left']['type']) {
-            $left = $expression['left']['left'];
-        } else {
-            $left = $expression['left'];
-        }
+        $left = 'list' === $expression['left']['type'] ? $expression['left']['left'] : $expression['left'];
 
         switch ($left['type']) {
             case 'array-access':
@@ -52,7 +50,7 @@ class IssetOperator extends AbstractOperator
                 $exprVariable->setNoisy(false);
 
                 $exprCompiledVariable = $exprVariable->compile($compilationContext);
-                if ('variable' != $exprCompiledVariable->getType() && 'array' != $exprCompiledVariable->getType()) {
+                if (!in_array($exprCompiledVariable->getType(), ['variable', 'array'], true)) {
                     throw new CompilerException('Expression type: '.$exprCompiledVariable->getType().' cannot be used as array', $left['left']);
                 }
 
@@ -67,7 +65,7 @@ class IssetOperator extends AbstractOperator
                         break;
                 }
 
-                if ('variable' == $variable->getType()) {
+                if ('variable' === $variable->getType()) {
                     if ($variable->hasDifferentDynamicType(['undefined', 'array', 'null'])) {
                         $compilationContext->logger->warning(
                             'Possible attempt to use non array in isset operator',
@@ -88,6 +86,7 @@ class IssetOperator extends AbstractOperator
                         return $compilationContext->backend->arrayIsset($variable, $resolvedExpr, $left['right'], $compilationContext);
 
                     case 'variable':
+                    case 'mixed':
                         $indexVariable = $compilationContext->symbolTable->getVariableForRead($resolvedExpr->getCode(), $compilationContext, $left['right']);
 
                         return $compilationContext->backend->arrayIsset($variable, $indexVariable, $left['right'], $compilationContext);
@@ -106,12 +105,12 @@ class IssetOperator extends AbstractOperator
                 $exprVariable->setReadOnly(true);
 
                 $exprCompiledVariable = $exprVariable->compile($compilationContext);
-                if ('variable' != $exprCompiledVariable->getType()) {
+                if ('variable' !== $exprCompiledVariable->getType()) {
                     throw new CompilerException('Expression type: '.$exprCompiledVariable->getType().' cannot be used as object', $left['left']);
                 }
 
                 $variable = $compilationContext->symbolTable->getVariableForRead($exprCompiledVariable->getCode(), $compilationContext, $left['left']);
-                if ('variable' != $variable->getType()) {
+                if ('variable' !== $variable->getType()) {
                     throw new CompilerException('Variable type: '.$variable->getType().' cannot be used as object', $left['left']);
                 }
 
@@ -123,7 +122,7 @@ class IssetOperator extends AbstractOperator
                 }
                 $variableCode = $compilationContext->backend->getVariableCode($variable);
 
-                if ('property-access' == $left['type']) {
+                if ('property-access' === $left['type']) {
                     return $compilationContext->backend->propertyIsset($variable, $left['right']['value'], $compilationContext);
                 }
 
@@ -137,6 +136,7 @@ class IssetOperator extends AbstractOperator
                         $indexVariable = $compilationContext->symbolTable->getVariableForRead($resolvedExpr->getCode(), $compilationContext, $left['right']);
                         switch ($indexVariable->getType()) {
                             case 'variable':
+                            case 'mixed':
                             case 'string':
                                 $indexVariableCode = $compilationContext->backend->getVariableCode($indexVariable);
 

--- a/Library/Operators/Other/IssetOperator.php
+++ b/Library/Operators/Other/IssetOperator.php
@@ -29,7 +29,7 @@ class IssetOperator extends AbstractOperator
     /**
      * Compiles an 'isset' operator.
      *
-     * @param array $expression
+     * @param array              $expression
      * @param CompilationContext $compilationContext
      *
      * @return CompiledExpression

--- a/Library/Passes/StaticTypeInference.php
+++ b/Library/Passes/StaticTypeInference.php
@@ -14,16 +14,14 @@ namespace Zephir\Passes;
 use Zephir\StatementsBlock;
 
 /**
- * StaticTypeInference.
- *
  * This pass try to infer typing on dynamic variables so the compiler
  * can replace them by low level types automatically
  */
 class StaticTypeInference
 {
-    protected $variables = [];
+    protected array $variables = [];
 
-    protected $infered = [];
+    protected array $infered = [];
 
     /**
      * Do the compilation pass.
@@ -156,6 +154,7 @@ class StaticTypeInference
                 break;
 
             case 'variable':
+            case 'mixed':
                 $this->variables[$variable] = 'undefined';
                 break;
 
@@ -243,11 +242,7 @@ class StaticTypeInference
     {
         if (isset($expression['parameters'])) {
             foreach ($expression['parameters'] as $parameter) {
-                if ('variable' == $parameter['parameter']['type']) {
-                    //$this->markVariable($parameter['value']);
-                } else {
-                    $this->passExpression($parameter['parameter']);
-                }
+                $this->passExpression($parameter['parameter']);
             }
         }
     }
@@ -255,11 +250,7 @@ class StaticTypeInference
     public function passArray(array $expression)
     {
         foreach ($expression['left'] as $item) {
-            if ('variable' == $item['value']['type']) {
-                //$this->markVariable($item['value']['value'], 'dynamical');
-            } else {
-                $this->passExpression($item['value']);
-            }
+            $this->passExpression($item['value']);
         }
     }
 
@@ -267,11 +258,7 @@ class StaticTypeInference
     {
         if (isset($expression['parameters'])) {
             foreach ($expression['parameters'] as $parameter) {
-                if ('variable' == $parameter['parameter']['type']) {
-                    //$this->markVariable($parameter['value'], 'dynamical');
-                } else {
-                    $this->passExpression($parameter['parameter']);
-                }
+                $this->passExpression($parameter['parameter']);
             }
         }
     }
@@ -561,27 +548,17 @@ class StaticTypeInference
                     }
                     break;
 
+                case 'throw':
                 case 'return':
                     if (isset($statement['expr'])) {
                         $this->passExpression($statement['expr']);
                     }
                     break;
 
+                case 'try-catch':
                 case 'loop':
                     if (isset($statement['statements'])) {
                         $this->passStatementBlock($statement['statements']);
-                    }
-                    break;
-
-                case 'try-catch':
-                    if (isset($statement['statements'])) {
-                        $this->passStatementBlock($statement['statements']);
-                    }
-                    break;
-
-                case 'throw':
-                    if (isset($statement['expr'])) {
-                        $this->passExpression($statement['expr']);
                     }
                     break;
 

--- a/Library/Statements/Let/ObjectDynamicProperty.php
+++ b/Library/Statements/Let/ObjectDynamicProperty.php
@@ -27,7 +27,7 @@ class ObjectDynamicProperty
     /**
      * Compiles foo->{x} = {expr}.
      *
-     * @param string $variable
+     * @param string             $variable
      * @param ZephirVariable     $symbolVariable
      * @param CompiledExpression $resolvedExpr
      * @param CompilationContext $compilationContext

--- a/Library/Statements/Let/ObjectDynamicProperty.php
+++ b/Library/Statements/Let/ObjectDynamicProperty.php
@@ -9,16 +9,17 @@
  * the LICENSE file that was distributed with this source code.
  */
 
+declare(strict_types=1);
+
 namespace Zephir\Statements\Let;
 
+use Exception;
 use Zephir\CompilationContext;
 use Zephir\CompiledExpression;
 use Zephir\Exception\CompilerException;
 use Zephir\Variable as ZephirVariable;
 
 /**
- * ObjectDynamicProperty.
- *
  * Updates object properties dynamically
  */
 class ObjectDynamicProperty
@@ -26,16 +27,16 @@ class ObjectDynamicProperty
     /**
      * Compiles foo->{x} = {expr}.
      *
-     * @param string             $variable
+     * @param string $variable
      * @param ZephirVariable     $symbolVariable
      * @param CompiledExpression $resolvedExpr
      * @param CompilationContext $compilationContext
      * @param array              $statement
      *
      * @throws CompilerException
-     * @throws \Exception
+     * @throws Exception
      */
-    public function assign($variable, ZephirVariable $symbolVariable, CompiledExpression $resolvedExpr, CompilationContext $compilationContext, array $statement)
+    public function assign(string $variable, ZephirVariable $symbolVariable, CompiledExpression $resolvedExpr, CompilationContext $compilationContext, array $statement)
     {
         if (!$symbolVariable->isInitialized()) {
             throw new CompilerException("Cannot mutate variable '".$variable."' because it is not initialized", $statement);
@@ -48,7 +49,7 @@ class ObjectDynamicProperty
         $propertyName = $statement['property'];
 
         $propertyVariable = $compilationContext->symbolTable->getVariableForRead($propertyName, $compilationContext, $statement);
-        if ($propertyVariable->isNotVariableAndString()) {
+        if ($propertyVariable->isNotVariableAndMixedAndString()) {
             throw new CompilerException("Cannot use variable type '".$propertyVariable->getType()."' to update object property", $statement);
         }
 
@@ -64,7 +65,7 @@ class ObjectDynamicProperty
             throw new CompilerException('Cannot use non-initialized variable as an object', $statement);
         }
 
-        /*
+        /**
          * Trying to use a non-object dynamic variable as object
          */
         if ($symbolVariable->hasDifferentDynamicType(['undefined', 'object', 'null'])) {
@@ -98,14 +99,14 @@ class ObjectDynamicProperty
                 break;
 
             case 'bool':
-                $value = null;
                 if ('1' == $resolvedExpr->getBooleanCode()) {
                     $value = 'true';
                 } elseif ('0' == $resolvedExpr->getBooleanCode()) {
                     $value = 'false';
                 } else {
-                    throw new \Exception('?');
+                    throw new Exception('?');
                 }
+
                 $compilationContext->backend->updateProperty($symbolVariable, $propertyVariableName, $value, $compilationContext);
                 break;
 
@@ -143,6 +144,7 @@ class ObjectDynamicProperty
                     case 'string':
                     case 'variable':
                     case 'array':
+                    case 'mixed':
                         $compilationContext->backend->updateProperty($symbolVariable, $propertyVariable, $resolvedExpr, $compilationContext);
                         if ($symbolVariable->isTemporal()) {
                             $symbolVariable->setIdle(true);

--- a/Library/Statements/Let/ObjectDynamicStringProperty.php
+++ b/Library/Statements/Let/ObjectDynamicStringProperty.php
@@ -26,7 +26,7 @@ class ObjectDynamicStringProperty
     /**
      * Compiles foo->{"x"} = {expr}.
      *
-     * @param string $variable
+     * @param string             $variable
      * @param ZephirVariable     $symbolVariable
      * @param CompiledExpression $resolvedExpr
      * @param CompilationContext $compilationContext

--- a/Library/Statements/Let/ObjectDynamicStringProperty.php
+++ b/Library/Statements/Let/ObjectDynamicStringProperty.php
@@ -9,6 +9,8 @@
  * the LICENSE file that was distributed with this source code.
  */
 
+declare(strict_types=1);
+
 namespace Zephir\Statements\Let;
 
 use Zephir\CompilationContext;
@@ -17,8 +19,6 @@ use Zephir\Exception\CompilerException;
 use Zephir\Variable as ZephirVariable;
 
 /**
- * ObjectDynamicProperty.
- *
  * Updates object properties dynamically
  */
 class ObjectDynamicStringProperty
@@ -26,7 +26,7 @@ class ObjectDynamicStringProperty
     /**
      * Compiles foo->{"x"} = {expr}.
      *
-     * @param string             $variable
+     * @param string $variable
      * @param ZephirVariable     $symbolVariable
      * @param CompiledExpression $resolvedExpr
      * @param CompilationContext $compilationContext
@@ -35,13 +35,13 @@ class ObjectDynamicStringProperty
      * @throws CompilerException
      * @throws \Exception
      */
-    public function assign($variable, ZephirVariable $symbolVariable, CompiledExpression $resolvedExpr, CompilationContext $compilationContext, array $statement)
+    public function assign(string $variable, ZephirVariable $symbolVariable, CompiledExpression $resolvedExpr, CompilationContext $compilationContext, array $statement)
     {
         if (!$symbolVariable->isInitialized()) {
             throw new CompilerException("Cannot mutate variable '".$variable."' because it is not initialized", $statement);
         }
 
-        if ('variable' != $symbolVariable->getType()) {
+        if ('variable' !== $symbolVariable->getType()) {
             throw new CompilerException("Variable type '".$symbolVariable->getType()."' cannot be used as object", $statement);
         }
 
@@ -63,7 +63,7 @@ class ObjectDynamicStringProperty
             throw new CompilerException('Cannot use non-initialized variable as an object', $statement);
         }
 
-        /*
+        /**
          * Trying to use a non-object dynamic variable as object
          */
         if ($symbolVariable->hasDifferentDynamicType(['undefined', 'object', 'null'])) {

--- a/Library/Statements/Let/ObjectProperty.php
+++ b/Library/Statements/Let/ObjectProperty.php
@@ -9,6 +9,8 @@
  * the LICENSE file that was distributed with this source code.
  */
 
+declare(strict_types=1);
+
 namespace Zephir\Statements\Let;
 
 use Zephir\CompilationContext as Context;
@@ -17,8 +19,6 @@ use Zephir\Exception\CompilerException as Exception;
 use Zephir\Variable as ZephirVariable;
 
 /**
- * ObjectProperty.
- *
  * Adds/Updates an array index
  */
 class ObjectProperty
@@ -285,6 +285,7 @@ class ObjectProperty
                     case 'array':
                     case 'string':
                     case 'variable':
+                    case 'mixed':
                         $context->backend->updateProperty($symbolVariable, $propertyName, $variableVariable, $context);
                         if ($symbolVariable->isTemporal()) {
                             $symbolVariable->setIdle(true);

--- a/Library/Statements/UnsetStatement.php
+++ b/Library/Statements/UnsetStatement.php
@@ -13,6 +13,7 @@ declare(strict_types=1);
 
 namespace Zephir\Statements;
 
+use ReflectionException;
 use Zephir\CompilationContext;
 use Zephir\Exception;
 use Zephir\Exception\CompilerException;
@@ -29,6 +30,7 @@ class UnsetStatement extends StatementAbstract
      * @param CompilationContext $compilationContext
      *
      * @throws Exception
+     * @throws ReflectionException
      */
     public function compile(CompilationContext $compilationContext): void
     {
@@ -61,6 +63,7 @@ class UnsetStatement extends StatementAbstract
                 break;
 
             case 'property-access':
+            case 'property-dinamic-access':
                 $expr = new Expression($expression['left']);
                 $expr->setReadOnly(true);
                 $exprVar = $expr->compile($compilationContext);
@@ -71,9 +74,6 @@ class UnsetStatement extends StatementAbstract
                 $compilationContext->codePrinter->output('zephir_unset_property('.$variableCode.', "'.$expression['right']['value'].'");');
 
                 return;
-
-            case 'property-dynamic-access':
-                //TODO: fix it
 
             default:
                 throw new CompilerException('Cannot use expression type: '.$expression['type'].' in "unset"', $expression);
@@ -94,12 +94,13 @@ class UnsetStatement extends StatementAbstract
     }
 
     /**
-     * @param array              $expression
+     * @param array $expression
      * @param CompilationContext $compilationContext
      *
      * @return CompilationContext
      *
      * @throws Exception
+     * @throws ReflectionException
      */
     private function generateUnsetPropertyFromObject(array $expression, CompilationContext $compilationContext): CompilationContext
     {

--- a/Library/Statements/UnsetStatement.php
+++ b/Library/Statements/UnsetStatement.php
@@ -94,7 +94,7 @@ class UnsetStatement extends StatementAbstract
     }
 
     /**
-     * @param array $expression
+     * @param array              $expression
      * @param CompilationContext $compilationContext
      *
      * @return CompilationContext

--- a/Library/Types/ArrayType.php
+++ b/Library/Types/ArrayType.php
@@ -18,8 +18,6 @@ use Zephir\Expression\Builder\BuilderFactory;
 use Zephir\Types;
 
 /**
- * ArrayType.
- *
  * Defines methods of the built-in array type
  */
 class ArrayType extends AbstractType

--- a/Library/Types/DoubleType.php
+++ b/Library/Types/DoubleType.php
@@ -14,8 +14,6 @@ namespace Zephir\Types;
 use Zephir\Types;
 
 /**
- * DoubleType.
- *
  * Defines methods of the built-in double type
  */
 class DoubleType extends AbstractType

--- a/Library/Types/IntType.php
+++ b/Library/Types/IntType.php
@@ -14,8 +14,6 @@ namespace Zephir\Types;
 use Zephir\Types;
 
 /**
- * IntType.
- *
  * Defines methods of the built-in int type
  */
 class IntType extends AbstractType

--- a/Library/Types/IstringType.php
+++ b/Library/Types/IstringType.php
@@ -14,8 +14,6 @@ namespace Zephir\Types;
 use Zephir\Types;
 
 /**
- * IstringType.
- *
  * Encapsulates built-in methods for the "istring" type
  */
 class IstringType extends StringType

--- a/Library/Types/StringType.php
+++ b/Library/Types/StringType.php
@@ -14,8 +14,6 @@ namespace Zephir\Types;
 use Zephir\Types;
 
 /**
- * StringType.
- *
  * Encapsulates built-in methods for the "string" type
  *
  * TODO: explode, join, replace

--- a/Library/Variable.php
+++ b/Library/Variable.php
@@ -1073,6 +1073,16 @@ class Variable implements TypeAwareInterface
     }
 
     /**
+     * Shortcut is type mixed?
+     *
+     * @return bool
+     */
+    public function isMixed(): bool
+    {
+        return 'mixed' === $this->type;
+    }
+
+    /**
      * Shortcut is type bool?
      *
      * @return bool
@@ -1140,6 +1150,16 @@ class Variable implements TypeAwareInterface
     public function isNotVariableAndString()
     {
         return !$this->isVariable() && !$this->isString();
+    }
+
+    /**
+     * Shortcut is type variable or mixed or string?
+     *
+     * @return bool
+     */
+    public function isNotVariableAndMixedAndString()
+    {
+        return !$this->isVariable() && !$this->isMixed() && !$this->isString();
     }
 
     /**

--- a/stub/arrayaccessarr.zep
+++ b/stub/arrayaccessarr.zep
@@ -1,6 +1,6 @@
 namespace Stub;
 
-class ArrayAccessObj implements \ArrayAccess
+class ArrayAccessArr implements \ArrayAccess
 {
 	protected test;
 

--- a/stub/arrayaccessobj.zep
+++ b/stub/arrayaccessobj.zep
@@ -37,6 +37,8 @@ class ArrayAccessObj implements \ArrayAccess
 	public function offsetUnset(mixed offset) -> void
 	{
 	    var obj = this->test;
+
+	    // It is not possible to unset dynamic property via Zend API.
 	    let obj = (array)obj;
 
         unset obj[offset];

--- a/stub/arrayaccessobj.zep
+++ b/stub/arrayaccessobj.zep
@@ -1,0 +1,51 @@
+namespace Stub;
+
+class ArrayAccessObj implements \ArrayAccess
+{
+	protected test;
+
+	public function __construct()
+	{
+	    var obj;
+
+	    let obj = new \stdClass();
+        let obj->key1 = "val1";
+        let obj->key2 = "val2";
+        let obj->key3 = "val3";
+
+        let this->test = obj;
+	}
+
+	public function offsetSet(mixed offset, mixed value) -> void
+	{
+	    var obj = this->test;
+
+	    if !is_null(offset) {
+	        let obj->{offset} = value;
+	    }
+
+	    let this->test = obj;
+	}
+
+	public function offsetExists(mixed offset) -> bool
+	{
+	    var obj = this->test;
+
+		return isset obj->{offset};
+	}
+
+	public function offsetUnset(mixed offset) -> void
+	{
+	    var obj = this->test;
+	    let obj = (array)obj;
+
+        unset obj[offset];
+
+        let this->test = (object)obj;
+	}
+
+	public function offsetGet(mixed offset) -> mixed
+	{
+		return isset this->test->{offset} ? this->test->{offset} : null;
+	}
+}

--- a/stub/arrayaccesstest.zep
+++ b/stub/arrayaccesstest.zep
@@ -17,7 +17,7 @@ class ArrayAccessTest
 	{
 		var arr;
 
-		let arr = new ArrayAccessObj();
+		let arr = new ArrayAccessArr();
 		return isset arr["one"];
 	}
 
@@ -25,7 +25,7 @@ class ArrayAccessTest
 	{
 		var arr;
 
-		let arr = new ArrayAccessObj();
+		let arr = new ArrayAccessArr();
 		return arr["two"];
 	}
 

--- a/tests/Extension/ArrayAccessObjTest.php
+++ b/tests/Extension/ArrayAccessObjTest.php
@@ -1,0 +1,52 @@
+<?php
+
+/**
+ * This file is part of the Zephir.
+ *
+ * (c) Phalcon Team <team@zephir-lang.com>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Extension;
+
+use PHPUnit\Framework\TestCase;
+use Stub\ArrayAccessObj;
+
+final class ArrayAccessObjTest extends TestCase
+{
+    public function testOffsetSetAndOffsetExists(): void
+    {
+        $class = new ArrayAccessObj();
+
+        $class->offsetSet('newKey', 'new value');
+        $this->assertTrue($class->offsetExists('newKey'));
+    }
+
+    public function testOffsetGet(): void
+    {
+        $class = new ArrayAccessObj();
+
+        $class->offsetSet('newKey', 'new value');
+        $this->assertSame('new value', $class->offsetGet('newKey'));
+
+        $class->offsetSet('newKey', 1);
+        $this->assertSame(1, $class->offsetGet('newKey'));
+
+        $class->offsetSet('boolean', true);
+        $this->assertSame(true, $class->offsetGet('boolean'));
+    }
+
+    public function testOffsetUnset(): void
+    {
+        $class = new ArrayAccessObj();
+
+        $class->offsetSet('newKey', 'new value');
+        $this->assertTrue($class->offsetExists('newKey'));
+        $class->offsetUnset('newKey');
+        $this->assertFalse($class->offsetExists('newKey'));
+    }
+}

--- a/tests/Extension/ArrayAccessObjTest.php
+++ b/tests/Extension/ArrayAccessObjTest.php
@@ -37,7 +37,7 @@ final class ArrayAccessObjTest extends TestCase
         $this->assertSame(1, $class->offsetGet('newKey'));
 
         $class->offsetSet('boolean', true);
-        $this->assertSame(true, $class->offsetGet('boolean'));
+        $this->assertTrue($class->offsetGet('boolean'));
     }
 
     public function testOffsetUnset(): void

